src/main/java/net/snowflake/ingest/streaming/internal/MemoryInfoProviderFromRuntime.java

```diff
@@ -4,20 +4,51 @@
 
 package net.snowflake.ingest.streaming.internal;
 
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
 /** Reads memory information from JVM runtime */
 public class MemoryInfoProviderFromRuntime implements MemoryInfoProvider {
-  @Override
-  public long getMaxMemory() {
-    return Runtime.getRuntime().maxMemory();
+  private final long maxMemory;
+  private volatile long totalFreeMemory;
+  private final ScheduledExecutorService executorService;
+  private static final long FREE_MEMORY_UPDATE_INTERVAL_MS = 1000;
+  private static final MemoryInfoProviderFromRuntime INSTANCE =
```

Comment:
@sfc-gh-tjones note that I have now made it a singleton since we don't want multiple instances of this in the same JVM all polling for free memory. One downside is that we cannot make the polling interval configurable easily anymore but that does not seem too bad to do it once a second.